responsibleai/responsibleai/rai_insights/rai_base_insights.py

```diff
@@ -0,0 +1,317 @@
+        return configs
+
+    def get(self):
+        """List information about each of the managers.
+
+        :return: Information about each of the managers.
+        :rtype: dict
+        """
+        configs = {}
+        for manager in self._managers:
+            configs[manager.name] = manager.get()
+        return configs
+
+    @abstractmethod
+    def get_data(self):
+        """Get all data as RAIInsightsData object
+
+        :return: Model Analysis Data
+        :rtype: RAIInsightsData
+        """
+        pass
+
+    @abstractmethod
+    def _get_dataset(self):
+        pass
+
+    def _write_to_file(self, file_path, content):
+        """Save the string content to the given file path.
+        :param file_path: The file path to save the content to.
+        :type file_path: str
+        :param content: The string content to save.
+        :type content: str
+        """
+        with open(file_path, 'w') as file:
+            file.write(content)
+
+    @abstractmethod
+    def _save_predictions(self, path):
+        """Save the predict() and predict_proba() output.
+
+        :param path: The directory path to save the RAIInsights to.
+        :type path: str
+        """
+        pass
+
+    def _save_data(self, path):
+        """Save the copy of raw data (train and test sets) and
+           their related metadata.
+
+        :param path: The directory path to save the RAIBaseInsights to.
+        :type path: str
+        """
+        data_directory = Path(path) / _DATA
+        data_directory.mkdir(parents=True, exist_ok=True)
+        dtypes = self.train.dtypes.astype(str).to_dict()
+        self._write_to_file(data_directory /
+                            (_TRAIN + _DTYPES + _JSON_EXTENSION),
+                            json.dumps(dtypes))
+        self._write_to_file(data_directory / (_TRAIN + _JSON_EXTENSION),
+                            self.train.to_json(orient='split'))
+
+        dtypes = self.test.dtypes.astype(str).to_dict()
+        self._write_to_file(data_directory /
+                            (_TEST + _DTYPES + _JSON_EXTENSION),
+                            json.dumps(dtypes))
+        self._write_to_file(data_directory / (_TEST + _JSON_EXTENSION),
+                            self.test.to_json(orient='split'))
+
+    @abstractmethod
+    def _save_metadata(self, path):
+        """Save the metadata like target column, categorical features,
+           task type and the classes (if any).
+
+        :param path: The directory path to save the RAIBaseInsights to.
+        :type path: str
+        """
+        pass
+
+    def _save_model(self, path):
+        """Save the model and the serializer (if any).
+
+        :param path: The directory path to save the RAIInsights to.
+        :type path: str
+        """
+        top_dir = Path(path)
+        if self._serializer is not None:
+            # save the model
+            self._serializer.save(self.model, top_dir)
```

Comment:
The serializer is assumed to handle the case of the model being None?

Reply:
The custom serializer should handle the model the user passed in.  Having said that, if the user passed in None for the model, it would be a bit weird for them to pass in a custom model serializer.  Perhaps we can throw in the _validate_model_analysis_input_parameters in this case?  I created an issue to track this:

https://github.com/microsoft/responsible-ai-toolbox/issues/1319

We can continue the conversation on that issue perhaps.